Fix Autopilot by calculating movement in the present time

TaterToes opened the pull request:
- Refer to #19930 and #25354 for issues.

Created a pull request since I believe its in the state of feedback and debugging plus cleanup

This implementation overhauls Autopilot’s logic to eliminate both the stutter, difficult tapping. and the teleporting between HitObjects. The cursor is given a velocity calculated based on avaliable time and the distance between the cursor to the next hit object.

It also fixes some other issues given by the TODO's.

Anyways, heres a video of the implementation below.

https://github.com/user-attachments/assets/9f6ad58f-ea56-4c7f-9d45-ece3108fa0db

TaterToes committed:
Remove replay frames from Autopilot

TaterToes committed:
Variable change + calculation clean up

TaterToes committed:
Merge branch 'ppy:master' into autopilot-branch

TaterToes committed:
Implement automatic spinning

TaterToes committed:
Add a comment

TaterToes committed:
name changes + redo spinner calculation

TaterToes committed:
CodeFactor Fix #1

TaterToes committed:
spinner rate fix

peppy committed:
Initial code quality pass

peppy commented on osu.Game.Rulesets.Osu/Mods/OsuModAutopilot.cs:
```diff
@@ -35,28 +36,132 @@ public class OsuModAutopilot : Mod, IUpdatableByPlayfield, IApplicableToDrawable
+        // Values to make cursor movement seem more natural.
+        // Assuming people cant tap within (MinStart) ms within an interval.
+        // Most likely needs a review.
+        private const double min_start = 20;
+        private const double min_end = 5;
```
These names are nondescript and the comment doesn't really help to describe what they are doing/for.

TaterToes replied:
I apologize, but let me explain this first before I update it.

My main worry about this PR was so that the cursor doesn't seemingly teleport from one hit object to the other. What min_start does is that if the currentTime is equal to the start of the hit window minus the min_start (avaliableTime is equal to min_start at this time), we gradually approach 1 ms of avaliableTime so that we can mostly avoid seemingly cursor teleportation until the end of the hitwindow minus the min_end. 

The problem is that if someone tapped faster then the time interval of min_start, the cursor wouldnt be on the hitcircle on time and would be considered as a miss.

peppy commented on osu.Game.Rulesets.Osu/Mods/OsuModAutopilot.cs:
```diff
@@ -35,38 +37,150 @@ public class OsuModAutopilot : Mod, IUpdatableByPlayfield, IApplicableToDrawable
-            // TODO: Implement the functionality to automatically spin spinners
+            double hitWindowStart = start - mehWindow - min_start;
+            double hitWindowEnd = start + mehWindow - min_end;
+            double availableTime = currentTime >= hitWindowStart
```
Shouldn't this just be a constant value rather than constantly changing every frame? This plus `moveTowards` are so obfuscated in weird logic that I don't see the reason to it all. This is either due to the math in this PR being "wahtever works lol", or my knowledge of math lacking, but either way I can't follow.

TaterToes replied:
I remember adding a comment here about what it does, but I might've accidentally deleted it.

My main worry was that I wanted the cursor to enter the HitObject at the time where HitWindow starts. Giving a constant value means that after we judged the last HitObject, it would take a constant value of time until the cursor reaches the HitCircle.

The other fact is that in moveTowards, the distance between the cursor and the hitobject is also changing. If we kept one of these constants, they would either accelerate or deaccelerate. I wanted a constant velocity, so either we need a constant value for both avaliableTime and distance, or each value has to change proportionally to each other to accurately calculate somewhat constant velocity. (well atleast until currentTime reaches hitWindow - minStart, that's basically the final guard against seemingly cursor teleportation)

TaterToes replied:
also the logic in moveTowards is that if both currentTime and distance are constantly updating. Velocity is calculated by how far it would go in 1 ms, and then multiply by how much time has passed between frames in ms.

TaterToes replied:
oh yeah i also forgot that sometimes hitobjects also moves

peppy reviewed (commented):
I've fixed up most of the code quality issues, and it seems to play well, but will need some further code revision.

TaterToes commented:
the meh window isn't calculated correctly when its a slider

TaterToes committed:
comments + fixes + name changes

TaterToes committed:
forgot to save more comments

TaterToes committed:
place hasReplayLoaded.Value in spots that forcefully moves cursor.

TaterToes committed:
oops.

TaterToes committed:
small changes to remove unnecessary code

I believe that guard is unnecessary

TaterToes committed:
Accidental statement that sets always to true on replay.

peppy commented:
> [place hasReplayLoaded.Value in spots that forcefully moves cursor.](https://github.com/ppy/osu/pull/32929/commits/1580098e32f2ae581f483d14fb82d2cc68063942)

why?

TaterToes commented:
> > [place hasReplayLoaded.Value in spots that forcefully moves cursor.](https://github.com/ppy/osu/pull/32929/commits/1580098e32f2ae581f483d14fb82d2cc68063942)
> 
> why?

I was thinking of putting in applyCursor, but i'll explain.

Now this is the part where my knowledge lacks, but since for some reason when I was watching the replay. The spinner wasn't being automatically spun.

TaterToes committed:
Stop mouse position updating by putting it in applyCursor

TaterToes committed:
Recalculate AvaliableTime

I realized that in the old calculation after it reaches a certain threshold, it would reach the center of the hit circle only at the end of the hitwindow minus an offset.

Now based on this calculation, we will reach the center of the HitObject and stay locked at the center after the first given avaliableTime of the current active HitObject. (ex. If it first calculated it to be 200 ms, its now expected to take approximately 200 ms)

I would say at lines 137 and 141 was only written to update saved_currentTime when we first enter the HitWindow of the current active HitObject. I would like a better way to do this.

I also reverted a commit that put hasReplayValue only in applyCursor.

Now, I would like this final part to be read. If we no longer want to worry about the mouse cursor movement, simply set avaliableTime and want a extremely simple calculation, we can simply set avaliableTime to Math.Max(hitWindowStart - currentTime, 1).

TaterToes committed:
code quality fixes + rearrangements

TaterToes committed:
code quality fix

TaterToes committed:
Accidental variable switch

TaterToes commented:
Now, I want to say that if we want to prioritize just have the cursor immediately on the HitObject when we are in the HitWindow, there would be no need for complicated code to calculate the time.

However, if we want the cursor to still act normal, this is my final product.

TaterToes committed:
Merge branch 'ppy:master' into autopilot-branch

TaterToes commented:
closing this down as I'm preparing a pull reques,t although similar, I feel like it's a better solution.
